\documentclass[a4paper]{article}

\usepackage[top=2cm, bottom=2cm, left=2.5cm, right=2.5cm]{geometry}
\usepackage{listings}
\usepackage{multirow}
\usepackage{xcolor}

\pagestyle{empty} % no page numbers
\setlength{\parindent}{0pt} % indenting the paragraphs looks silly here

% LISTINGS SETTING
\definecolor{hutton_lightgreen}{HTML}{C8F27F}

\lstset{ %
  backgroundcolor=\color{hutton_lightgreen},   % choose the background color; you must add \usepackage{color} or \usepackage{xcolor}
  basicstyle=\small\ttfamily,        % the size of the fonts that are used for the code
  breakatwhitespace=true,          % sets if automatic breaks should only happen at whitespace
  breaklines=true,                 % sets automatic line breaking
  captionpos=b,                    % sets the caption-position to bottom
  abovecaptionskip=-12pt,          % not using captions, so this avoids wasting vertical space
  belowcaptionskip=0pt,
  commentstyle=\color{red},    % comment style
  deletekeywords={...},            % if you want to delete keywords from the given language
  escapeinside={\%*}{*)},          % if you want to add LaTeX within your code
  extendedchars=true,              % lets you use non-ASCII characters; for 8-bits encodings only, does not work with UTF-8
  frame=single,                    % adds a frame around the code
  keepspaces=true,                 % keeps spaces in text, useful for keeping indentation of code (possibly needs columns=flexible)
  keywordstyle=\color{blue},       % keyword style
%  language=Octave,                 % the language of the code
  morekeywords={*,...},            % if you want to add more keywords to the set
  numbers=left,                    % where to put the line-numbers; possible values are (none, left, right)
  numbersep=5pt,                   % how far the line-numbers are from the code
  numberstyle=\tiny\color{gray}, % the style that is used for the line-numbers
  rulecolor=\color{black},         % if not set, the frame-color may be changed on line-breaks within not-black text (e.g. comments (green here))
  showspaces=false,                % show spaces everywhere adding particular underscores; it overrides 'showstringspaces'
  showstringspaces=false,          % underline spaces within strings only
  showtabs=false,                  % show tabs within strings adding particular underscores
  stepnumber=1,                    % the step between two line-numbers. If it's 1, each line will be numbered
  stringstyle=\color{violet},     % string literal style
  tabsize=4,                       % sets default tabsize to 4 spaces
  title=\lstname                   % show the filename of files included with \lstinputlisting; also try caption instead of title
}

\begin{document}

\section*{Introduction To Bioinformatics -- Command summary}

This sheet is just a summary of the commands you will be running, which are described in the slides.
You should have already created a copy of the data files under your home directory (shorthand tilde):
\begin{lstlisting}[language=bash]
$ cp -r /mnt/shared/public/training/introToBioinformatics/ ~/
\end{lstlisting}

Check the session's files exist with:
\begin{lstlisting}[language=bash]
$ ls ~/introToBioinformatics/data/
ex1_expression  ex2_blast  workshop
\end{lstlisting}

\section*{Part 1: Golden Rules of Bioinformatics}

\subsection*{Exercise 1 -- Expression data in R}

\begin{lstlisting}[language=bash]
$ cd ~/introToBioinformatics/data/ex1_expression/
$ R
\end{lstlisting}

\begin{lstlisting}[language=R]
> data = read.table("expnA.tab", sep="\t", header=TRUE)
> head(data)
\end{lstlisting}

\begin{lstlisting}[language=R]
> mean(data$gene1)
> mean(data$gene2)
> sd(data$gene1)
> sd(data$gene2)
> cor(data)
\end{lstlisting}

\begin{lstlisting}[language=R]
> plot(data)
\end{lstlisting}

\begin{lstlisting}[language=R]
> data = anscombe
> cor(data)[1:4,5:8]
\end{lstlisting}

\subsection*{Exercise 2 -- BLAST+ at the command line}

\begin{lstlisting}[language=bash]
$ cd ../ex2_blast
$ blastp -num_alignments 1 -num_descriptions 1 -query query.fasta -db dbA
BLASTP 2.2.30+

...
\end{lstlisting}

\section*{Part 2: BLAST}

\subsection*{BLAST+ at the command line}

Built in help:

\begin{lstlisting}[language=bash]
$ blastp -h
...
\end{lstlisting}

\pagebreak

Minimal example:

\begin{lstlisting}[language=bash]
$ cd ~/introToBioinformatics/data/ex2_blast/
$ blastp -query query.fasta -db dbA -out my_output.txt
\end{lstlisting}

Changing the output format:

\begin{lstlisting}[language=bash]
$ blastp -query query.fasta -db dbA -out my_output.txt
$ blastp -query query.fasta -db dbA -out my_output.xml -outfmt 5
$ blastp -query query.fasta -db dbA -out my_output.tab -outfmt 6
\end{lstlisting}

Setting the e-value threshold:

\begin{lstlisting}[language=bash]
$ blastp -query query.fasta -db dbA -out my_output.txt -evalue 1e-5
\end{lstlisting}

\section*{Part 3: Genome Annotation Workshop}

\subsection*{Gene Prediction}

\begin{lstlisting}[language=bash]
$ cd ~/introToBioinformatics/data/workshop/
$ ls *.fasta
chrA.fasta      chrB.fasta      chrC.fasta      chrD.fasta
$ head -n 3 chrA.fasta
$ head -n 3 chrA_contigs.gff
\end{lstlisting}

% \begin{lstlisting}[language=bash]
% $ art &
% \end{lstlisting}

\begin{lstlisting}[language=bash]
$ build-icm -r NC_004547.icm < NC_004547.ffn
$ glimmer3 -o 50 -g 110 -t 30 chrA.fasta NC_004547.icm chrA_glimmer3
$ head -n 4 chrA_glimmer3.predict
$ python glimmer_to_gff.py chrA_glimmer3.predict
$ head -n 3 chrA_glimmer3.gff
\end{lstlisting}

\begin{lstlisting}[language=bash]
$ prodigal -f gff -o chrA_prodigal.gff -i chrA.fasta
$ head -n 6 chrA_prodigal.gff
\end{lstlisting}

\subsection*{Genome Comparison (with BLAST+)}

\begin{lstlisting}[language=bash]
$ blastn -query chrA.fasta -subject NC_004547.fna -out chrA_megablast_Pba.tab -outfmt 6
$ head -n 3 chrA_megablast_Pba.tab
\end{lstlisting}

\begin{lstlisting}[language=bash]
$ blastn -query chrA.fasta -subject NC_004547.fna -out chrA_blastn_Pba.tab -outfmt 6 -task blastn
$ head -n 3 chrA_blastn_Pba.tab
\end{lstlisting}

\begin{lstlisting}[language=bash]
$ wc chrA_megablast_Pba.tab
$ wc chrA_blastn_Pba.tab
\end{lstlisting}

% \begin{lstlisting}[language=bash]
% $ act &
% \end{lstlisting}

\subsection*{Genome Comparison (with MUMmer)}

\begin{lstlisting}[language=bash]
$ cd ~/introToBioinformatics/data/workshop
$ mkdir nucmer_out
\end{lstlisting}

\begin{lstlisting}[language=bash]
$ nucmer --prefix=nucmer_out/chrA_NC_004547 chrA.fasta NC_004547.fna
$ delta-filter -q nucmer_out/chrA_NC_004547.delta > nucmer_out/chrA_NC_004547.filter
$ show-coords -rcl nucmer_out/chrA_NC_004547.filter > nucmer_out/chrA_NC_004547_filtered.coords
$ head nucmer_out/chrA_NC_004547_filtered.coords
\end{lstlisting}

\begin{lstlisting}[language=bash]
$ tail -n +6 nucmer_out/chrA_NC_004547_filtered.coords | awk '{print $7" "$10" "$1" "$2" "$12" "$4" "$5" "$13}' > chrA_mummer_NC_004547.crunch
$ head chrA_mummer_NC_004547.crunch
\end{lstlisting}

\end{document}
